### ISSUE TITLE
Leave original element when cloning

### ISSUE THREAD

**Issue by lewism36** [angular-ui-tree/angular-ui-tree#446]
Using the data-clone-enabled option, is there an easy way to leave the element in its place while the drag is in progress?  It seems to temporarily remove the element and drop it back in place once the drag/clone is complete

**Comment by Voles** [comment]
@lewism36 I don't quite catch your question. Using `clone-enabled`, the original element should not have been changed.

**Comment by lewism36** [comment]
Hi,


As a basic example I added the data-clone-enabled attribute to the trees in the ‘connected trees’ example.


When dragging a node from the left tree to the right, the left list reduces to 3 items until the node is dropped on the right list, then it appears again in the left list once it has been dropped on the right - I am looking for a way for the left list to remain intact at all times - i.e. the clone occurring as soon as the drag starts.


A further issue I found is that the underlying model object for the node is not actually cloned - it is just added to the other array - so the same object exists in both arrays - this is ok until it is dragged back again or the same item is cloned once more.  


This then causes a Duplicate Key in Repeater error in Angular because the same item exists twice in the same array.  I worked around this second issue by cloning the item myself in the dropped event but it would be useful if this happened automatically (using track by $index on the repeater caused strange issues when the trees had nested items).


hope that makes sense


Lewis









Sent from Windows Mail





From: Niels Dequeker
Sent: ‎Saturday‎, ‎2‎ ‎May‎ ‎2015 ‎16‎:‎14
To: angular-ui-tree/angular-ui-tree
Cc: lewism36





@lewism36 I don't quite catch your question. Using clone-enabled, the original element should not have been changed.

**Comment by fyodorvi** [comment]
I've got into exact same issues with duplicating upon dragging a clone. Could you please share your solution?

This should be done be default, though.

And item should never disappear from source when cloning is enabled, or at least there should be an option for that. That makes a lot sense.

**Comment by lewism36** [comment]
@fyodorvi  It's a bit messy, but I basically keep a local copy of my source array (it never needs to change after the controller initializes and populates it from the server), then in the drop event handler of the source tree I empty the array that the source tree is bound to, and fill it with clones of each item from the local array.  I then also update the event.source.nodeScope.$modelValue.$$hashKey to be something unique.  

I did first try cloning only the dropped item from the bound array, removing the original, then adding back the clone, but this didn't seem to work.

**Comment by fyodorvi** [comment]
@lewism36 I though about doing it your was, but yeah, it messy. So I edited ui-tree source. I modified dragInfo.appy function a bit, the last line that actually inserts node now inserts a copy of it if cloning is enabled:

Was:
```
this.parent.insertNode(this.index, nodeData) ;
```

Now:
```
this.parent.insertNode(this.index, this.source.$treeScope.cloneEnabled === true ? angular.copy(nodeData) : nodeData) ;
```

Which to my mind makes much more sense. 

To solve problem with disappearing item (just like you described) I modified dragStart function a bit: I don't remove source element if cloning is enabled (which is, again, reasonable for me), and don't add placeholder to source list:

Was:
```
   scope.$element.after(placeElm);
   scope.$element.after(hiddenPlaceElm);
   dragElm.append(scope.$element);
```

Now:
```
   if (scope.$treeScope.cloneEnabled) {
      dragElm.append(scope.$element.clone());
   } else {
      scope.$element.after(placeElm);
      dragElm.append(scope.$element);
   }
```

I hope this might help you.

**Comment by fyodorvi** [comment]
Last modification is not so smart, though. It assumes that you don't allow drop of clone on the source tree, so it should be treated as a must-have. Ideally there should be separate setting for that.

**Comment by lewism36** [comment]
thanks @fyodorvi - that's a good solution

**Comment by lewism36** [comment]
Actually I think there's a line missing from the else.  should it be:

if (scope.$treeScope.cloneEnabled) {
   dragElm.append(scope.$element.clone());
} else {
   scope.$element.after(placeElm);
   scope.$element.after(hiddenPlaceElm);
   dragElm.append(scope.$element);
}

**Comment by fyodorvi** [comment]
@lewism36, sorry, it should be like that:

```
scope.$element.after(hiddenPlaceElm);

 if (scope.$treeScope.cloneEnabled) {
      dragElm.append(scope.$element.clone());
   } else {
      scope.$element.after(placeElm);
      dragElm.append(scope.$element);
   }
```

hiddenPlaceElm should always be there

**Comment by lewism36** [comment]
great, thanks

**Comment by honkskillet** [comment]
Why was this reopened?  If it was an accident, can OP close it again?

**Comment by lewism36** [comment]
Is it fixed?

**Comment by jacobscarter** [comment]
Closing this issue due to inactivity. See #815 for more information.
Please create a new issue is this still occurs with the current version of the component.
